### PR TITLE
Flow distance utility development

### PR DESCRIPTION
**Pull request by Giuseppecipolla95**
This begins to develop a few tests to demonstrate that flow length utility (#638) works as expected.

### CODE REVIEW

**Comment by kbarnhart** [comment]
@Giuseppecipolla95 I'm going to be away from internet for a bit starting tomorrow, so don't expect any responses from me for a few days.

**Comment by kbarnhart** [comment]
@nathanlyons would you be able to help me review @Giuseppecipolla95 's PR here and in the tutorials repository?
 
overall things look pretty good to me but I'd like to get another set of eyes from the dev team.

**Comment by nathanlyons** [comment]
@kbarnhart Yes, I will and I am looking forward to having this utility available. Thanks for asking. The EGU meeting is about over. I will review it likely this Sunday. Monday at the latest.

**Comment by kbarnhart** [comment]
thanks @nathanlyons! hope EGU has been good!

File: landlab/utils/tests/test_stream_length.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    #flow_length_expected = np.array ([[0,0,0,0],[0,1,0,0],[0,2,1,0],[0,3,2,0],[0,0,0,0]],dtype='float64')
+    fa.run_one_step()
+    # calculating flow length map
+    flow_length = calculate_stream_length(mg, add_to_grid=True, noclobber=False)
+    # test that the stream length utility works as expected
+    assert_array_equal(flow_length_expected, flow_length)
+
+
+def test_stream_length_raster_D_infinity():
+    """Test of stream length utility with a raster grid and D infinity."""
```
Comment (kbarnhart) [inline landlab/utils/tests/test_stream_length.py]: @Giuseppecipolla95 I think you should be able to create an expected stream length field for this. 

The issue you are having has to do with the fact that the grid you have created results in all of the nodes being assigned as flowing to themselves with DINF. I'll have to take a more close look and identify if this is a bug in D infinity or if it is behaving correctly. 

This will almost certainly happen after next Wednesday.

File: landlab/utils/tests/test_stream_length.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    flow_length = np.reshape(flow_length,mg.number_of_node_rows*mg.number_of_node_columns)
+    # modifying the flow_length_map because boundary and outlet nodes should not
+    # have stream_length value different from 0
+    flow_length[mg.boundary_nodes] = 0
+    outlet_id = 6
+    flow_length[outlet_id] = 0
+    mg.add_field('node','flow_length',flow_length)
+    # test that the stream length utility works as expected
+    assert_array_equal(flow_length_expected, flow_length)
+
```
Comment (kbarnhart) [inline landlab/utils/tests/test_stream_length.py]: This test appears to be correct to me.

File: landlab/utils/tests/test_stream_length.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    flow_length = np.reshape(flow_length,mg.number_of_node_rows*mg.number_of_node_columns)
+    # modifying the stream_length_map because boundary and outlet nodes should not
+    # have stream_length value different from 0
+    flow_length[mg.boundary_nodes] = 0
+    outlet_id = 6
+    flow_length[outlet_id] = 0
+    mg.add_field('node','flow_length',flow_length)
+    # test that the stream length utility works as expected
+    assert_array_equal(flow_length_expected, flow_length)
+
```
Comment (kbarnhart) [inline landlab/utils/tests/test_stream_length.py]: This test also looks correct

File: landlab/utils/tests/test_stream_length.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    fr = FlowAccumulator(hmg, flow_director = 'D4')
+    fr.run_one_step()
+    # calculating flow length map
+    flow_length = calculate_stream_length(hmg, add_to_grid=True, noclobber=False)
+    # modifying the stream_length_map because boundary and outlet nodes should not
+    # have stream_length value different from 0
+    flow_length[hmg.boundary_nodes] = 0
+    outlet_id = 4
+    flow_length[outlet_id] = 0
+    hmg.add_field('node','flow_length',flow_length)
```
Comment (kbarnhart) [inline landlab/utils/tests/test_stream_length.py]: You should not be making modifications like this within tests. This discrepancy has come up because you have not set any of the boundary nodes to be closed and so water is flowing out of the grid domain. 

you can do this with: 
```python
hmg.set_closed_nodes(hmg.boundary_nodes)
```

You'll want to
1) Delete these lines that modify the flow length after you calculate it
2) Either
a) Set the boundary nodes closed and use the calculated flow length you have OR
b) keep the boundary nodes open and recalculate the flow length.

File: landlab/utils/tests/test_stream_length.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    # considering flow directions calculated with MFD algorithm
+    flow_length_expected = np.array ([[0,0,0,0],[0,1,0,0],[0,math.sqrt(2),1,0],[0,1+math.sqrt(2),2,0],[0,0,0,0]],dtype='float64')
+    flow_length_expected = np.reshape (flow_length_expected,mg.number_of_node_rows*mg.number_of_node_columns)
+    #setting boundary conditions
+    mg.set_closed_boundaries_at_grid_edges(bottom_is_closed=True, left_is_closed=True, right_is_closed=True, top_is_closed=True)
+    # calculating flow directions with FlowAccumulator component
+    fa = FlowAccumulator(mg, 'topographic__elevation',flow_director='MFD',diagonals = True)
+    #If diagonals = False, the flow_length_expected becomes:
+    #flow_length_expected = np.array ([[0,0,0,0],[0,1,0,0],[0,2,1,0],[0,3,2,0],[0,0,0,0]],dtype='float64')
+    fa.run_one_step()
```
Comment (kbarnhart) [inline landlab/utils/tests/test_stream_length.py]: I'd make an additional test that uses `diagonals=False`

File: landlab/utils/tests/test_stream_length.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    # calculating flow length map
+    flow_length = calculate_stream_length(mg, add_to_grid=True, noclobber=False)
+    # test that the stream length utility works as expected
+    assert_array_equal(flow_length_expected, flow_length)
+
```
Comment (kbarnhart) [inline landlab/utils/tests/test_stream_length.py]: This test appears to me to be correct.

File: testlog.txt [hunk 7]
```diff
@@ -0,0 +1,528 @@
+/Users/barnhark/projects/landlab/landlab/landlab/utils/decorators.py:134: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
```
Comment (kbarnhart) [inline testlog.txt]: @Giuseppecipolla95 can you delete this file? I must have accidentally added it. My apologies. I didn't mean to. It was used for debugging purposes only.

File: landlab/utils/tests/test_stream_length.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    # test that the stream length utility will fail because of a ValueError
+    assert_raises(FieldError, calculate_stream_length, mg)
+
+
+def test_stream_length_regular_grid_d8():
+    """Test to demonstrate that stream_length utility works as expected with regular grids"""
+    # instantiate a model grid
+    mg = RasterModelGrid((5, 4), spacing=(1, 1))
+    # instantiate an elevation array
+    z = np.array ([[0,0,0,0],[0,21,10,0],[0,31,20,0],[0,32,30,0],[0,0,0,0]],dtype='float64')
```
Comment (nathanlyons) [inline landlab/utils/tests/test_stream_length.py]: PEP8 also indicates that a comma should be followed by a space. Also, no need for a space after a function name (e.g. np.array):
```python
z = np.array([[0, 0, 0, 0], [0, 21, 10, 0]...
```
Please adjust other instances incorrect spacing.

File: landlab/utils/tests/test_stream_length.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    flow_length_expected = np.array ([[0,0,0,0],[0,1,0,0],[0,math.sqrt(2),1,0],[0,1+math.sqrt(2),2,0],[0,0,0,0]],dtype='float64')
+    flow_length_expected = np.reshape (flow_length_expected,mg.number_of_node_rows*mg.number_of_node_columns)
+    #setting boundary conditions
+    mg.set_closed_boundaries_at_grid_edges(bottom_is_closed=True, left_is_closed=True, right_is_closed=True, top_is_closed=True)
+    # calculating flow directions with FlowAccumulator component
+    fr = FlowAccumulator(mg, flow_director = 'D8')
+    fr.run_one_step()
+    # calculating flow length map
+    flow_length = calculate_stream_length(mg, add_to_grid=True, noclobber=False)
+    flow_length = np.reshape(flow_length,mg.number_of_node_rows*mg.number_of_node_columns)
```
Comment (nathanlyons) [inline landlab/utils/tests/test_stream_length.py]: A space should be both before and after an operator, such as * .
```python
mg.number_of_node_rows * mg.number_of_node_columns
```

To handle the line length here:
```python
flow_length = np.reshape(flow_length, mg.number_of_node_rows *
                         mg.number_of_node_columns)
```

File: landlab/utils/tests/test_stream_length.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    # add the elevation field to the grid
+    mg.add_field('node','topographic__elevation', z)
+    # instantiate the expected flow_length array
+    # considering flow directions calculated with D8 algorithm
+    flow_length_expected = np.array ([[0,0,0,0],[0,1,0,0],[0,math.sqrt(2),1,0],[0,1+math.sqrt(2),2,0],[0,0,0,0]],dtype='float64')
+    flow_length_expected = np.reshape (flow_length_expected,mg.number_of_node_rows*mg.number_of_node_columns)
+    #setting boundary conditions
+    mg.set_closed_boundaries_at_grid_edges(bottom_is_closed=True, left_is_closed=True, right_is_closed=True, top_is_closed=True)
+    # calculating flow directions with FlowAccumulator component
+    fr = FlowAccumulator(mg, flow_director = 'D8')
```
Comment (nathanlyons) [inline landlab/utils/tests/test_stream_length.py]: Perhaps unexpectedly, no spaces are around equal signs for keyword arguments, although the equal side adjacent to the output (the output here is fr) does have spaces around it:
```python
fr = FlowAccumulator(mg, flow_director='D8')
```

File: landlab/utils/tests/test_stream_length.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    fd.run_one_step()
+    # test that the stream length utility will fail because of a ValueError
+    assert_raises(FieldError, calculate_stream_length, mg)
+
+
+def test_stream_length_regular_grid_d8():
+    """Test to demonstrate that stream_length utility works as expected with regular grids"""
+    # instantiate a model grid
+    mg = RasterModelGrid((5, 4), spacing=(1, 1))
+    # instantiate an elevation array
```
Comment (nathanlyons) [inline landlab/utils/tests/test_stream_length.py]: Please consider adding one blank line above a block of code that is similar in functionality. This helps readers to quickly scan lines. 

An example of what this looks like can be seen in the stream_length utility file in this pull request.

File: landlab/utils/tests/test_stream_length.py (hunk 1 above)
Comment (Giuseppecipolla95) [reply landlab/utils/tests/test_stream_length.py]: @kbarnhart  Ok, I'm waiting for your response.

**Comment by kbarnhart** [comment]
@Giuseppecipolla95 

I was able to look at DINF a bit this morning. 

I think the issue is a bug that relates to how I implemented the original D infinity algorithm in conjunction with how Landlab handles boundary conditions. Specifically, I first apply the standard D infinity algorithm to identify which two nodes flow should go to. If the steepest triangular facet leaving from the node flows entirely to closed nodes, the the node is marked as 'flows to self'. Instead I should mark out the closed nodes _before_ the D infinity algorithm runs. In your case, the edge nodes are all closed _and_ all have low elevations, so all the interior nodes were marked as 'flows to self'.

I'll be making an issue of this later today and dealing with this asap. 

Nice work making a test that identified this issue! Its always hard for us to test all edge cases to make sure that the code is working as expected. 

Importantly for you, by modifying the elevations you use in the D infinity test, D infinity will work as it should, and you should be able to make the stream length test. 

If you set up the test grid for D infinity with the following code, you should be in good shape. 

```python
mg = RasterModelGrid((5, 4), spacing=(1, 1))
# instantiate an elevation array
z = mg.x_of_node + 2.0 * mg.y_of_node
# add the elevation field to the grid
mg.add_field('node','topographic__elevation', z)
#setting boundary conditions

mg.set_closed_boundaries_at_grid_edges(bottom_is_closed=True, left_is_closed=True, right_is_closed=True, top_is_closed=True)
```

**Comment by nathanlyons** [comment]
Thanks for making my requested changes. I went ahead and approved, although I have one more suggestion that I also included in the tutorial review.

To me, stream length is applicable only to streams nodes (some decision must be made about which nodes are streams) and flow length is the length of simulated flow at every grid node. However, the stream_length utility calculates what I would call flow length. This is just a naming preference. I am not proposing a naming change.

Instead I suggest adjusting the utility documentation to make this clear. In stream_length.py, in the documentation between Parameters and Examples, the output can described:

> Returns
> --------
> stream__length : float ndarray
>     The length of flow at each node...

I wrote '...' because you are more qualified than I to finish this statement.

**Comment by Giuseppecipolla95** [comment]
@nathanlyons I've added the return in flow_length.py.
@kbarnhart I've completed the test regarding the Dinf algorithm, thanks for describing the bug.
Let me know if everything is ok.

File: landlab/utils/flow__distance.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+    if 'flow__receiver_node' not in grid.at_node:
+        raise FieldError("A 'flow__receiver_node' field is required at the "
+                         "nodes of the input grid.")
+    if 'flow__upstream_node_order' not in grid.at_node:
+        raise FieldError("A 'flow__upstream_node_order' field is required at the "
+                         "nodes of the input grid.")
+
+    # get an array of grid nodes
+    grid_nodes = grid.nodes.flatten()
+
```
Comment (kbarnhart) [inline landlab/utils/flow__distance.py]: @Giuseppecipolla95 
I think we are about ready to merge this, nice work!

I think that we can delete these two lines since `grid_nodes` is no longer used later. 

```python
# get an array of grid nodes
grid_nodes = grid.nodes.flatten()
``` 

You can either delete the lines yourself or add me as a collaborator on this fork. Either option is fine with me. 

Once that is done, we'll merge this branch and @nathanlyons will follow up with you regarding finishing the tutorial.

Comment (Giuseppecipolla95) [reply landlab/utils/flow__distance.py]: @kbarnhart I've added you as a collaborator on the fork, so you should be able to delete the two lines indicated by yourself on the comment.
Regarding the tutorial, I think it should be complete. 
@nathanlyons Please confirm it or let me know if anything else should be fixed.

**Comment by kbarnhart** [comment]
@mcflugen this is another case in which the travis tests actually failed but were summarized as successful. 

@Giuseppecipolla95 I removed that line of code and also worked on the tests so that they passed on my computer. 

If they actually pass on Travis and Apveyor, I'll merge this PR. 

@nathanlyons , I'll let you handle finalizing the tutorial.

**Comment by kbarnhart** [comment]
Looking at the tests, there are some that fail, but they are not related to these changes. So I've merged. 
Nice work @Giuseppecipolla95 !

**Comment by Giuseppecipolla95** [comment]
@kbarnhart Thanks for giving me the opportunity to collaborate!

**Comment by kbarnhart** [comment]
@Giuseppecipolla95 , would you be able to provide some feedback to us regarding our documentation? As someone who is outside the core development team, but interested and active in developing the Landlab code base, we would be very interested in your feedback. 

We would like to improve our documentation so that someone like you -- a motivated person who is interested in developing tools, utilities, or components in Landlab -- can make a good start on developing your tool based on information in the documentation. 

I would be most  interested in your feedback on [this page that discusses developing in Landlab](https://github.com/landlab/landlab/wiki/Develop-your-own-component) and any other page you found helpful, or not-quite-as-helpful-as-you-would-have wanted.

**Comment by Giuseppecipolla95** [comment]
@kbarnhart Is there a specific page where I can write my feedback? I'm developing a model, for my dissertation, that is able to obtain the flood hydrographs for all the Sicilian basins referred to a set of return times. The part of the extraction of watersheds and flow__distance maps from the DEM is done by using some of the Landlab components and utility. If you are interested to the use of Landlab in my code I'll publish it on Hydroshare ASAP before my degree, scheduled for July.

**Comment by kbarnhart** [comment]
@Giuseppecipolla95 you could put the feedback here, or create a new Issue with the feedback. Either one would be fine with us. 

Congratulations on nearing the end of your dissertation! We would love to see your work published on Hydroshare.
